src/analyzer/analyzerthread.h

```diff
@@ -118,7 +120,7 @@ class AnalyzerThread : public WorkerThread {
 
     mixxx::SampleBuffer m_sampleBuffer;
 
-    TrackPointer m_currentTrack;
+    std::optional<AnalyzerTrack> m_currentTrack;
```

Comment:
This is not a good fit for an optional, because the track inside inside can be still null even though the optional predicts that we have a value. You will have the same convenient API if you implement "has_value()" yourself without redundant data storage.

Reply:
I would rather like to do it the other way around, i.e. assert that the `TrackPointer` inside must not be null, since to me it's semantically the right thing: If there's no track, we probably don't want to attach any options either, so it feels cleaner to me to model `AnalyzerTrack` as a value that is never empty and then add optionality externally.

Reply:
Another question is if this can ever be null. We have recently added a std::not_null to our code base. 
I am afraid that this does not work with our smart track pointer. We need to have a look. 
In any case a not_null track pointer will help our code base a lot. (A topic for another PR ...)

Reply:
Regarding

> without redundant data storage.

IMO it's not the right place to micro-optimize: The extra byte that `std::optional` uses would be a cost that we pay per `AnalyzerThread` and not a very significant one.

Reply:
The discussion about the extra byte is mood, if we consider that the track is always valid. We can make this true, but lets do it in another PR to keep this one simple. 

Having two concurrent values the express an invalid state is a recipe of future issue that needs to be fixed IMHO. It is not that much a performance thing. For me the best solution for this PR is to add an own "has_value()" function.

Reply:
Yeah, let's move that to a separate PR. But I think it makes sense _semantically_ that an analyzer thread could either have a track or not. If we were to model optionality inside of `AnalyzerTrack` we might want to consider renaming that class, since it no longer represents a 'track' but an optional track, I'd really vote against that thought. Keeping `AnalyzerTrack` an individual, non-nullable unit that can be scheduled is IMO cleaner and doesn't mix in another concern (optionality).

Reply:
So we go for a "has_value()" member in this PR, right?

Reply:
I would really advise against it, since it goes against the design of this class (being a POD value type), doesn't buy us anything in terms of thread-safety, simplicity and barely makes a difference in space/performance, but if you feel strongly about this, then we can of course also do that.

The reason I believe `std::optional` makes sense here is that we already effectively did the same thing already, just implicitly: We used null pointers instead of C++'s optionals. The latter, however, are simultaneously more clear, harder to misuse, encode invariants in the type system and really don't have a big footprint.

Reply:
We need here a protecting against null pointer access. Using another value for it is errophrone.  

Look at our code base and the qt code base. The are many cases with an invalid state internally and a API function to access that. Wrapping that by an optional make the code harder to use and more complex. 

Optionals are a perfect fit for types like booleans without an invalid value. But please don't use it for pointers.

Reply:
I have read up on the documentation of `gsl::not_null` and it seems to be supported to use it for smart pointers. So I've wrapped `AnalyzerTrack`'s `TrackPointer` in `gsl::not_null` and updated the tests as necessary. In `TrackAnalysisScheduler` (where we usually construct `AnalyzerTrack`s) we already perform a null-check, so constructing the `not_null` should be safe.

I believe this is the cleanest solution, since illegal states should now be unrepresentable and we can compose optionality separately on top (of a combined not-null track-with-options). It also improves local reasoning since optionality as an effect is more visible than having to worry about an internal pointer that may or may not be null.

Let me know if this works for you.